Contact support configure chat widget

ravishanker opened the pull request:
Configures chat interface to match the designs as much as possible
- Provides essential options to bot like id, firstMessage, color etc....
- Removes top `X Close` button
- Removes top `Reset` button
- Removes top Header
- Adds `Contact support` button

Fixes #18900 
This also fixes part of #18828 

| Start | Conversation |
|--------|--------|
|![Screenshot_20230811_034120](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/2291d512-3767-451c-8e5d-d88bc9d7b91f) | ![Screenshot_20230811_034138](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/0ac8ef09-c2fc-4e76-828d-f4eab56dfbc9)| 


1. Launch Jetpack app
2. Go to `Me` (Tap on Avatar in the bottom nav)
3. Tap `Debug Settings`
4. Find `contact_support` under `Remote features` tab, and enable it, return to Help screen
5. Tap on Help -> Contact Support
6. `Verify` it launches Support Bot as shown above
7. `Send a message...` in the text field
8. `Verify` bot responds with an answer
9. `Verify` there's a `Contact support` button below response
10. `Tap` on the `Contact support`
11. `Verify` the chat windows is closed, and returns to Help screen
12. `Observe` in AS Logcat Chat history is logged
13. `Tap` on `Contact support` again on `Help` screen
14. `Tap` on back arrow in the header, and ensure it returns to previous `Help` screen

>**Note**
> Further enhancements could be done in the next PRs

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually

3. What automated tests I added (or what prevented me from doing so)
Existing unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Create JsObject.kt

ravishanker committed:
Update SupportWebViewActivity.kt

ravishanker committed:
Update HelpActivity.kt

ravishanker committed:
Update support_chat_widget.js

ravishanker committed:
Update support_chat_widget.html

ravishanker committed:
Update HelpViewModel.kt

ravishanker committed:
Update SupportWebViewClient.kt

ravishanker committed:
Refactor and cleanup

ravishanker committed:
Use string resources

Add string resources to aid translation

ravishanker committed:
reset conversation

irfano reviewed (changes requested):
I tested the PR but I don't see the bot as the screenshot at the first opening. And when I tap the pre-written questions, it returns error. This was working before, I guess we broke something in this PR.

<img width="300" alt="Screenshot 2023-08-10 at 15 55 53" src="https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/70439099-3098-458c-ac5f-54eaab886d09">

irfano commented on WordPress/src/jetpack/assets/support_chat_widget.html:
```diff
@@ -1,21 +1,9 @@
-        }, 200);
-    </script>
-</body>
+    <head>
+        <script type="text/javascript" src="/assets/bot_widget.js"></script>
+    </head>
+    <body>
+        <script type="text/javascript" src="/assets/support_chat_widget.js"></script>
+    </body>
```
Why are there two scripts for the bot here, `bot_widget.js` and `support_chat_widget.js`? I guess you were trying something, and we'll remove one of these, right?

ravishanker committed:
Update support_chat_widget.js

decodeURIComponent

ravishanker commented:
> I tested the PR but I don't see the bot as the screenshot at the first opening. And when I tap the pre-written questions, it returns error. This was working before, I guess we broke something in this PR.
> 

Looks like broke when I updated to use Strings.xml. Needed to decodeURIComponent.  Fixed now.  Thanks.

irfano commented:
Thanks, Ravi! The pre-written questions are working now. But the style is still l broken at the first launch. Do you have any guesses as to what could be the reason?

https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/83a78cae-ebe0-46b5-9e57-e1fb9499b888

ravishanker commented:
> Thanks, Ravi! The pre-written questions are working now. But the style is still l broken at the first launch. Do you have any guesses as to what could be the reason?
> 

What style are you referring to?  

Chat widget at first launch was appearing before as well sometimes.  Still need to figure it out, and improve that part as we progress along.

This PR also removes the top header, and close buttons to make it look like chat. (This can be easily tweaked later).  

Pushed another update improving the first time, and subsequent load 80c5df511657b7a5a1c4f119437e18f42904de03

ravishanker committed:
Update support_chat_widget.js

Improve first time loading of bot in full screen

staskus commented on WordPress/src/jetpack/assets/support_chat_widget.js:
```diff
@@ -1,31 +1,109 @@
+                    thinking: "Thinking..."
+                }, // Override all the default labels for your own language.
+                questions: [
+                    decodeURIComponent(urlParams.questionOne),
```
As I understand, the suggested questions can be set from the DocsBot dashboard. Should we rely on the dashboard and not set local versions of our questions? I think there're benefits of them being dynamic and their updates not reliant on app updates. WDYT?

**Update**

I didn't take into account that language can only be configured from the dashboard. In that case, for questions to appear in languages other than English we need to set them in the app.

ravishanker replied:
correct

staskus commented on WordPress/src/jetpack/assets/support_chat_widget.js:
```diff
@@ -1,31 +1,109 @@
+                description: "Ask me anything about using Jetpack mobile app.",
+                color: "#9dd977",
+                supportLink: "#",
+                labels: {
```
~~We can also pass "language" as an option.~~

~~I need to do some testing, maybe if we pass a locale identifier here, we would get strings translated ~~"for free"~~ in the chat interface if the language is supported by the bot. Otherwise, the chat would appear in English. Maybe it would help avoid some issues if we get these strings translated but the bot actually doesn't support the language?~~

**Update**

I misunderstood the documentation. We can only change the language from the bot's settings page. Then yes, for bot to appear localized we need to localize those strings as well, then rely on internals of the bot being able to handle the language.

staskus commented on WordPress/src/jetpack/assets/support_chat_widget.js:
```diff
@@ -1,31 +1,109 @@
+
+    function resetConversation() {
+        // reset button was in top header
+        localStorage.removeItem("docsbot_chat_history");
```
It doesn't always reset conversation for me:


https://github.com/wordpress-mobile/WordPress-Android/assets/4062343/72d65d1e-941c-44b2-b4a4-09f636ca24e5

staskus replied:
I tested, it looks like if we call `resetConversation();` before `openDocsBot` when this issue does not happen

ravishanker replied:
It was intentional not to reset the conversation always immediately, if they want to continue sometimes from where they left off.

We could always tweak it as we learn more

staskus replied:
The problem I noticed, is that after reopening it and clicking "Contact support", history only contains new objects and doesn't contain history before closing/reopening. That's why I think it's a good idea to `resetConversation()` every time.

staskus commented on WordPress/src/jetpack/assets/support_chat_widget.js:
```diff
@@ -1,31 +1,109 @@
+        }).then(() => {
+            // Safely do stuff here after the widget is loaded.
+            let timerId = setInterval(() => {
+                openDocsBot(); // wait for init
```
This method does not always work, perhaps a `200` interval is not always sufficient. During testing on both iOS and Android, I have observed that sometimes the chat appears as expected, while other times we need to click on the chat icon.

ravishanker replied:
Probably, need to look at with the help of a JS expert or DocsBot people to open straight away instead of a widget, and click/open mechanism.

staskus replied:
True. I'm looking at it but it's hard to say why it doesn't work sometimes. `.init` fulfills a promise when `docsbotai-root` element is found. It makes sense that at that moment we should be able to open the chat. I tried other approaches, such as observing `#docsbotai-root` myself. It seems to be more reliable although I cannot find arguments why it's more reliable.

staskus reviewed (commented):
✅ Colors
✅ Removes top X Close button
✅ Removes top Reset button
✅ Removes top Header
✅ Adds Contact support button
🟡 Sometimes the chat doesn't open automatically
🟡 Sometimes the history is not reset

staskus commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -2279,6 +2279,14 @@
+    <string name="contact_support_get_support">Contact support</string>
+    <string name="contact_support_suggestions">Not sure what to ask?</string>
+    <string name="contact_support_question_one">How do I change home page?</string>
+    <string name="contact_support_question_two">Account or billing issues?</string>
```
Yes, we need to make sure that the answer to this question is tapping "Contact Support" below the message.

<img width="486" alt="image" src="https://github.com/wordpress-mobile/WordPress-Android/assets/4062343/a4a40e5a-302a-4b8d-a7de-044d6d28f703">

irfano commented:
> What style are you referring to?

I'm referring the top header, close button and not starting from the chat screen. 

> Pushed another update improving the first time, and subsequent load https://github.com/wordpress-mobile/WordPress-Android/commit/80c5df511657b7a5a1c4f119437e18f42904de03

The issue still persists in my tests. Every time I kill the app and then relaunch it, the chat screen doesn't start immediately, and the top header and close button remain visible. 

We can proceed with merging this PR and address this problem in future PRs. But I recommend removing `Fixes #18828` from the description to ensure this fix is not overlooked.

staskus commented on WordPress/src/jetpack/assets/support_chat_widget.js:
```diff
@@ -1,31 +1,109 @@
+    }
+
+    function openDocsBot() {
+        const widget = document.querySelector("#docsbotai-root").shadowRoot.querySelector("a.floating-button");
```
Would it be enough to just call `DocsBotAI.open();`?

ravishanker replied:
That's what was there before, wasn't working, so tried mimicking click on a widget if it appears.  Give it a try, with only open(), see if it works.  Better if we can figure out a way to avoid the widget altogether!

staskus replied:
Got it. Makes sense. I'm trying some approaches locally as well. I will let you know if anything works.

ravishanker commented:
> We can proceed with merging this PR and address this problem in future PRs

Yes, please.  You're going to need this for submit ticket PR.

irfano reviewed (approved):
I'm approving the PR, despite the fact that the widget still requires further improvements. This is necessary in order to divide the tasks and prepare the `trunk` ready for additional PRs, such as #18902. 

`Fixes #18828 → This also fixes part of #18828`
I have updated the PR description to ensure that #18828 will not be automatically closed upon merging, as #18828 still requires additional improvements.

@staskus, thank you for your valuable comments on the code. Could we proceed with merging this PR and then address the remaining comments in a subsequent PR?

staskus commented:
> Could we proceed with merging this PR and then address the remaining comments in a subsequent PR?

@irfano yes, we can leave improvements for future PRs.
